stdlib/SparseArrays/test/sparse.jl

```diff
@@ -2764,4 +2764,57 @@ end
     @test adjoint(B)*adjoint(complex.(A)) ≈ adjoint(Array(B)) * adjoint(Array(complex.(A)))
 end
 
+# Check issue 33332.
+# The == operator is supposed to ignore comparing zero()
+# The = operator (setindex!) is supposed to not add zero() elements
+@testset "Sparse of structure redefining zero(). Check == and = operators" begin
+    # Define a struct redefining the zero()
+    struct MP{T} <: Number λ::T end
+    Base.zero(::Type{MP{T}}) where T = MP(typemin(T))
+    Base.zero(x::MP{T})      where T = zero(typeof(x))
+    Base.promote_rule(::Type{MP{T}}, ::Type{U}) where {T, U} = MP{T}
+    Base.convert(::MP{T}, x) where T = MP(T(x))
+
+    # Check if Julia is confused between mp0 (zero(MP) = -Inf) and 0 (zero(Float) = 0.0)
+    mp0 = zero(MP{Float64})
+
+    # A contains 'fake' zeros and B is empty
+    A = sparse([1, 2], [1, 2], [MP(0.0), 0.0])
+    B = spzeros(MP{Float64}, 2,2)
+    @test nonzeros(A) == [MP(0.0); 0.0]
+    @test nonzeros(B) == []
+    @test (A == B) == false
+    @test (A != B) == true
+
+    # AA contains newly defined zeros and BB is empty
+    AA = sparse([1, 2], [1, 2], [mp0, mp0])
+    BB = spzeros(MP{Float64}, 2,2)
+    @test nonzeros(AA) == [MP(-Inf); -Inf]
+    @test nonzeros(BB) == []
+    @test (AA == BB) == true
+    @test (AA != BB) == false
+
+    # Check if 'classic' zero is not broken.
+    AAA = sparse([1, 2], [1, 2], [0.0, 0.0])
+    BBB = spzeros(Float64, 2,2)
+    @test nonzeros(AAA) == [0.0; 0.0]
+    @test nonzeros(BBB) == []
+    @test (AAA == BBB) == true
+    @test (AAA != BBB) == false
+
+    # Inserting newly defined zero. Check if B is still empty
+    B[1,1] = mp0
+    @test nonzeros(B) == []
```

Comment:
I cannot changed it because isempty(B) returns false while nzval is empty. Seems because isempty(B) == length(B) which is 4 in our case (like a dense matrix). Maybe a fix would be isempty(B) == length(B.nzval)? What do you think? I think this can be fixed in another commit!

Reply:
It's just a detail, but personally, I find `== []` a bit ugly. You could test for `@test isempty(nonzeros(B))`, for example.

Reply:
Yes I can change. but what do you think isempty(B) should (length(B.nzval) == 0). Should I open a ticket ?

Reply:
changed!

Reply:
No, that works as supposed:
```julia
isempty(a::AbstractArray) = (length(a) == 0)
length(t::AbstractArray) = (@_inline_meta; prod(size(t)))
```
The length of `B` is 4, as it doesn't depend on the values contained in `B`. You can check with `@which fun(args)` which method is used, and check for its definition in the source code. I'd recommend to do that before filing issues, just to make sure to understand what the intended behavior is.

Reply:
I did `@which` but this is very confusing: there is no comment so we do not know if it's the dimension of the 'container' of if the matrix has no element.

Reply:
PS: according to `?`  isempty = Determine whether a collection is empty (has no elements). So I think I'm right: should `isempty(s::SparseMatrixCSC) = (length(s.nzval) == 0)` which is also compatible with `isempty([])`

Reply:
What do you mean? I found the above two definitions with `@which`, and that tells me that `isempty` of an array eventually checks whether the product of its dimensions is zero, which is clearly not the case for your `B`.

Reply:
Well, I won't stop you from filing issues, but as you can see from the copied code above, this would be against the current generic behavior of `isempty` applied to matrices. If you know you're dealing with a sparse matrix, what's the problem of asking `isempty(nonzeros(B))` if that's exactly what you're interested in?

Reply:
when you say `the product of its dimensions is zero` you're right point of view of the code but the documentation says something else `Determine whether a collection is empty (has no elements)`. In our case B is empty: zero() elements are not stored so empty. No ?

Reply:
No. Generally, a matrix or an array is not empty just because it contains only zeros. That's because generally, a zero value may be an important piece of information, which is why in normal arrays we do store zeros. A matrix is empty if, according to its size, it cannot possibly contain any elements. This concept of emptiness is applied to all matrix/array types.

Now, in the case of sparse matrices, the values are stored in a vector that you retrieve via `nonzeros`. So, under the "constraint" that zero elements are not stored, that vector is going to be empty. So, what you're really interested in is whether `nonzeros(B)` is empty. And that's what you yourself started testing even in your first commit, so I think everything is consistent here. Maybe should move this discussion to http://discourse.julialang.org.